File: internal/execution/invocation_abi_conformance_test.go
package execution

import (
	"encoding/base64"
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

// TestInvocationABIConformanceCorpus runs the shared Go↔NEMO corpus
// against the Go parser. NEMO runs the same corpus from
// nemo/test/invocation-abi-conformance.test.ts — both runtimes must
// accept or reject each raw wire request identically, and every
// rejection must include the corpus's rule-level error phrase.
func TestInvocationABIConformanceCorpus(t *testing.T) {
	type vector struct {
		Name          string `json:"name"`
		Wire          string `json:"wire"`
		WireB64       string `json:"wire_b64"`
		Expected      string `json:"expected"`
		ErrorContains string `json:"error_contains"`
	}
	var corpus struct {
		Vectors []vector `json:"vectors"`
	}
	data, err := os.ReadFile(filepath.Join("testdata", "invocation-abi-conformance", "vectors.json"))
	if err != nil {
		t.Fatalf("read corpus: %v", err)
	}
	if err := json.Unmarshal(data, &corpus); err != nil {
		t.Fatalf("parse corpus: %v", err)
	}
	if len(corpus.Vectors) == 0 {
		t.Fatal("conformance corpus is empty")
	}
	for _, v := range corpus.Vectors {
		t.Run(v.Name, func(t *testing.T) {
			wire := []byte(v.Wire)
			if v.WireB64 != "" {
				decoded, err := base64.StdEncoding.DecodeString(v.WireB64)
				if err != nil {
					t.Fatalf("decode wire_b64: %v", err)
				}
				wire = decoded
			}
			_, err := parseInvocationRequest(wire)
			switch v.Expected {
			case "accept":
				if err != nil {
					t.Fatalf("expected acceptance, got error: %v", err)
				}
			case "reject":
				if err == nil {
					t.Fatal("expected rejection, got acceptance")
				}
				if v.ErrorContains != "" && !strings.Contains(err.Error(), v.ErrorContains) {
					t.Fatalf("error %q does not contain %q", err, v.ErrorContains)
				}
			default:
				t.Fatalf("unknown expectation %q", v.Expected)
			}
		})
	}
}

File: nemo/test/invocation-abi-conformance.test.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { describe, expect, it } from "vitest";

import { validateInvocationRequest } from "../contracts/invocation-abi";

const corpusPath = join(
  dirname(fileURLToPath(import.meta.url)),
  "../../internal/execution/testdata/invocation-abi-conformance/vectors.json",
);

interface ConformanceVector {
  name: string;
  wire?: string;
  wire_b64?: string;
  expected: "accept" | "reject";
  error_contains?: string;
}

const corpus = JSON.parse(readFileSync(corpusPath, "utf-8")) as {
  description: string;
  vectors: ConformanceVector[];
};

// The same corpus runs against the Go parser in
// internal/execution/invocation_abi_conformance_test.go: both runtimes
// must accept or reject each raw wire request identically.
describe("capability invocation ABI conformance (Go↔NEMO)", () => {
  it("corpus is non-empty", () => {
    expect(corpus.vectors.length).toBeGreaterThan(0);
  });

  for (const vector of corpus.vectors) {
    it(`${vector.expected}s ${vector.name}`, () => {
      const bytes = vector.wire_b64
        ? new Uint8Array(Buffer.from(vector.wire_b64, "base64"))
        : new Uint8Array(Buffer.from(vector.wire ?? "", "utf-8"));
      const result = validateInvocationRequest(bytes);
      if (vector.expected === "accept") {
        expect(result.ok, result.ok ? "" : result.error).toBe(true);
        return;
      }
      expect(result.ok).toBe(false);
      if (!result.ok && vector.error_contains) {
        expect(result.error).toContain(vector.error_contains);
      }
    });
  }
});
